File: tests/testrun.py
#!/usr/bin/env python
from basetest import BaseTest
import os, sys, subprocess
import unittest

sys.path.insert(0, '..')

# (testing command support imports zeroinstall.injector._runenv in a sub-process)
os.environ['PYTHONPATH'] = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

from zeroinstall import SafeException

mydir = os.path.abspath(os.path.dirname(__file__))
local_0launch = os.path.join(mydir, '..', 'build', 'ocaml', '0launch')
arglist = os.path.join(mydir, 'runnable', 'ArgList.xml')
runnable = os.path.join(mydir, 'runnable', 'Runnable.xml')
runexec = os.path.join(mydir, 'runnable', 'RunExec.xml')
recursive_runner = os.path.join(mydir, 'runnable', 'RecursiveRunner.xml')
command_feed = os.path.join(mydir, 'Command.xml')
package_selections = os.path.join(mydir, 'package-selection.xml')

class TestRun(BaseTest):
	def testRunnable(self):
		child = subprocess.Popen([local_0launch, '--', runnable, 'user-arg'], stdout = subprocess.PIPE, universal_newlines = True)
		stdout, _ = child.communicate()
		assert 'Runner: script=A test script: args=command-arg -- user-arg' in stdout, stdout

	def testCommandBindings(self):
		if 'SELF_COMMAND' in os.environ:
			del os.environ['SELF_COMMAND']

		out, err = self.run_ocaml(['run', '--main=runnable/go.sh', '-wenv #', command_feed])
		assert not err, err
		assert 'LOCAL=' in out, out
		assert 'SELF_COMMAND=' in out, out

	def testAbsMain(self):
		out, err = self.run_ocaml(['run', '--dry-run', '--main=runnable/runner', command_feed])
		assert '[dry-run] would execute: ././runnable/runner' in out, out
		assert not err, err

		out, err = self.run_ocaml(['run', '--main=runnable/not-there', command_feed])
		assert not out, out
		assert 'not-there' in err, err

	def testBadMain(self):
		out, err = self.run_ocaml(['run', '--dry-run', '--command=', command_feed])
		assert "Exit status: 1" in err, err
		assert "Can't run: no command specified!" in err, err

		out, err = self.run_ocaml(['run', '--dry-run', '--command=', '--main=relpath', command_feed])
		assert "Exit status: 1" in err, err
		assert "Can't use a relative replacement main (relpath) when there is no original one!" in err, err

	def testArgs(self):
		out, err = self.run_ocaml(['run', '--dry-run', runnable])
		assert not err, err
		assert 'runner-arg' in out, out

	def testArgList(self):
		out, err = self.run_ocaml(['run', '--dry-run', arglist])
		assert not err, err
		assert 'arg-for-runner -X ra1 -X ra2' in out, out
		assert 'command-arg ca1 ca2' in out, out

	def testWrapper(self):
		out, err = self.run_ocaml(['run', '-wecho', '--dry-run', runnable])
		assert not err, err
		assert '/bin/sh -c echo "$@"' in out, out
		assert 'runner-arg' in out, out
		assert 'script' in out, out

	def testRecursive(self):
		child = subprocess.Popen([local_0launch, '--', recursive_runner, 'user-arg'], stdout = subprocess.PIPE, universal_newlines = True)
		stdout, _ = child.communicate()
		assert 'Runner: script=A test script: args=command-arg -- arg-for-runnable recursive-arg -- user-arg' in stdout, stdout

	def testExecutable(self):
		child = subprocess.Popen([local_0launch, '--', runexec, 'user-arg-run'], stdout = subprocess.PIPE, universal_newlines = True)
		stdout, _ = child.communicate()
		assert 'Runner: script=A test script: args=foo-arg -- var user-arg-run' in stdout, stdout
		assert 'Runner: script=A test script: args=command-arg -- path user-arg-run' in stdout, stdout

	def testRunPackage(self):
		if 'TEST' in os.environ:
			del os.environ['TEST']
		child = subprocess.Popen([local_0launch, '--wrapper', 'echo $TEST #', '--', package_selections], stdout = subprocess.PIPE, universal_newlines = True)
		stdout, _ = child.communicate()
		assert stdout.strip() == 'OK', stdout
	
if __name__ == '__main__':
	unittest.main()

File: src/tests/test_run.ml
(* Copyright (C) 2013, Thomas Leonard
 * See the README file for details, or visit http://0install.net.
 *)

open Support.Common
open OUnit

let assert_str_equal = Fake_system.assert_str_equal
let assert_contains = Fake_system.assert_contains

let arglist = Fake_system.test_data "runnable" +/ "ArgList.xml"
let runnable = Fake_system.test_data "runnable" +/ "Runnable.xml"
let runexec = Fake_system.test_data "runnable" +/ "RunExec.xml"
let recursive_runner = Fake_system.test_data "runnable" +/ "RecursiveRunner.xml"
let command_feed = Fake_system.test_data "Command.xml"
let package_selections = Fake_system.test_data "package-selection.xml"


let run_0install (fake_system:Fake_system.fake_system) args =
  fake_system#set_spawn_handler (Some Fake_system.real_system#create_process);
  Fake_system.fake_log#reset;
  fake_system#set_argv @@ Array.of_list (Fake_system.test_0install :: args);
  Fake_system.capture_stdout ~include_stderr:false (fun () ->
    try
      let stdout = Format.std_formatter in
      Main.main ~stdout (fake_system : Fake_system.fake_system :> system);
      assert false
    with Fake_system.Did_exec -> ()
  )

let suite = "run">::: [
  "runnable">:: Fake_system.with_fake_config (fun (_config, fake_system) ->
    skip_if on_windows "No /bin/sh";
    let out = run_0install fake_system ["run"; "--"; runnable; "user-arg"] in
    assert_str_equal "Runner: script=A test script: args=command-arg -- user-arg\n" out;
  );

  "command-bindings">:: Fake_system.with_fake_config (fun (_config, fake_system) ->
    skip_if on_windows "No /bin/sh";
    let out = run_0install fake_system ["run"; "--main=runnable/go.sh"; "-wenv #"; command_feed] in
    assert_contains "LOCAL=" out;
    assert_contains "SELF_COMMAND=" out
  );

  "abs-main">:: Fake_system.with_fake_config (fun (_config, fake_system) ->
    let out = Test_0install.run_0install fake_system ["run"; "--dry-run"; "--main=runnable/runner"; command_feed] in
    assert_contains "[dry-run] would execute:" out;
    if on_windows then assert_contains ".\\runnable/runner" out
    else assert_contains "./runnable/runner" out;

    Fake_system.assert_raises_safe ".*not-there" (lazy (
      Test_0install.run_0install fake_system ["run"; "--main=runnable/not-there"; command_feed] |> ignore;
    ));
  );

  "bad-main">:: Fake_system.with_fake_config (fun (_config, fake_system) ->
    Fake_system.assert_raises_safe "Can't run: no command specified!" (lazy (
      Test_0install.run_0install fake_system ~exit:1 ["run"; "--dry-run"; "--command="; command_feed] |> ignore
    ));

    Fake_system.assert_raises_safe "Can't use a relative replacement main (relpath) when there is no original one!" (lazy (
      Test_0install.run_0install fake_system ["run"; "--dry-run"; "--command="; "--main=relpath"; command_feed] |> ignore
    ));
  );

  "args">:: Fake_system.with_fake_config (fun (_config, fake_system) ->
    let out = Test_0install.run_0install fake_system ["run"; "--dry-run"; runnable] in
    assert_contains "runner-arg" out
  );

  "arg-list">:: Fake_system.with_fake_config (fun (_config, fake_system) ->
    let out = Test_0install.run_0install fake_system ["run"; "--dry-run"; arglist] in
    assert_contains "arg-for-runner -X ra1 -X ra2" out;
    assert_contains "command-arg ca1 ca2" out
  );

  "wrapper">:: Fake_system.with_fake_config (fun (_config, fake_system) ->
    let out = Test_0install.run_0install fake_system ["run"; "-wecho"; "--dry-run"; runnable] in
    assert_contains "/bin/sh -c echo \"$@\"" out;
    assert_contains "runner-arg" out;
    assert_contains "script" out
  );

  "recursive">:: Fake_system.with_fake_config (fun (_config, fake_system) ->
    skip_if on_windows "No /bin/sh";
    let out = run_0install fake_system ["run"; "--"; recursive_runner; "user-arg"] in
    assert_contains "Runner: script=A test script: args=command-arg -- arg-for-runnable recursive-arg -- user-arg" out
  );

  "executable">:: Fake_system.with_fake_config (fun (_config, fake_system) ->
    skip_if on_windows "No /bin/sh";
    let out = run_0install fake_system ["run"; "--"; runexec; "user-arg-run"] in
    assert_contains "Runner: script=A test script: args=foo-arg -- var user-arg-run" out;
    assert_contains "Runner: script=A test script: args=command-arg -- path user-arg-run" out
  );

  "run-package">:: Fake_system.with_fake_config (fun (_config, fake_system) ->
    skip_if on_windows "No /bin/sh";
    let out = run_0install fake_system ["run"; "--wrapper"; "echo $TEST #"; "--"; package_selections] in
    assert_str_equal "OK" (String.trim out)
  );
]
